File: FMELibrary/NationParser.cs
namespace FMELibrary
{
    public class NationParser
    {
        /// <summary>
        /// File path
        /// </summary>
        public string FilePath { get; set; }

        /// <summary>
        /// File header, 8 bytes
        /// </summary>
        public byte[] Header { get; set; }

        /// <summary>
        /// Item count
        /// </summary>
        public short Count { get; set; }

        /// <summary>
        /// List of all items
        /// </summary>
        public List<Nation> Items { get; set; }

        private NationParser(string path, BinaryReader reader)
        {
            FilePath = path;
            Header = reader.ReadBytes(8);
            Count = reader.ReadInt16();
            Items = [];
        }

        public static async Task<NationParser> Load(string path)
        {
            using var fs = File.OpenRead(path);
            using var ms = new MemoryStream();
            fs.CopyTo(ms);
            ms.Position = 0;

            using var reader = new BinaryReader(ms);
            var parser = new NationParser(path, reader);

            await Task.Run(() =>
            {
                while (ms.Position < ms.Length)
                {
                    var item = new Nation(reader);
                    parser.Items.Add(item);
                }
            });

            return parser;
        }

        public byte[] ToBytes()
        {
            using var stream = new MemoryStream();
            using var writer = new BinaryWriter(stream);

            writer.Write(Header);
            writer.Write((short)Items.Count);
            foreach (var item in Items)
            {
                item.Write(writer);
            }

            return stream.ToArray();
        }

        /// <summary>
        /// Save data back to file path
        /// </summary>
        /// <param name="filepath"></param>
        /// <returns></returns>
        public async Task Save(string? filepath = null)
        {
            var bytes = ToBytes();
            await File.WriteAllBytesAsync(filepath ?? FilePath, bytes);
        }
    }
}

File: FMELibrary/NationParser.ts
import { promises as fs } from 'fs';
import { Nation } from './Nation';

/**
 * Parses and manages nation data from binary files.
 */
export class NationParser {
  /** File path */
    filePath: string = '';

    /** File header, 8 bytes */
    header: Buffer = Buffer.alloc(0);

    /** Item count */
    count: number = 0;

    /** List of all items */
    items: Nation[] = [];

    /**
     * Initializes a new instance of the NationParser class.
     * @param path - The file path of the source data
     * @param buffer - The buffer containing the nation data
     */
    private constructor(path: string, buffer: Buffer) {
this.filePath = path;
  this.header = buffer.slice(0, 8);
        this.count = buffer.readInt16LE(8);
        this.items = [];
    }

    /**
     * Asynchronously loads nation data from the specified file path.
     * @param path - The file path to load nation data from
     * @returns A promise that resolves to the loaded NationParser instance
     */
    static async load(path: string): Promise<NationParser> {
        const buffer = await fs.readFile(path);
   const parser = new NationParser(path, buffer);

        let offset = 10; // Skip header (8) and count (2)

     while (offset < buffer.length) {
      const result = Nation.fromBuffer(buffer, offset);
            parser.items.push(result.nation);
            offset = result.offset;

         // Debug output
   console.log(`Loaded Nation: ${result.nation.name}`);
        }

        return parser;
    }

    /**
     * Converts the nation data to a buffer for serialization.
     * @returns A buffer representing the serialized nation data
     */
    toBuffer(): Buffer {
        const buffers: Buffer[] = [];
      
        buffers.push(this.header);
        
        const countBuffer = Buffer.alloc(2);
      countBuffer.writeInt16LE(this.items.length, 0);
        buffers.push(countBuffer);

        for (const item of this.items) {
            buffers.push(item.toBuffer());
        }

        return Buffer.concat(buffers);
    }

/**
     * Asynchronously saves data back to file path.
     * @param filepath - Optional file path. If null, saves to the original file path
     * @returns A promise that resolves when the save operation is complete
     */
    async save(filepath?: string): Promise<void> {
        const buffer = this.toBuffer();
  await fs.writeFile(filepath ?? this.filePath, buffer);
    }
}
